File: go/indicators/geraldappel/percentagepriceoscillator/percentagepriceoscillator.go
package percentagepriceoscillator

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/common/exponentialmovingaverage"
	"zpano/indicators/common/simplemovingaverage"
	"zpano/indicators/core"
)

// lineUpdater is an interface for indicators that accept a single scalar and return a value.
type lineUpdater interface {
	Update(float64) float64
	IsPrimed() bool
}

// PercentagePriceOscillator is Gerald Appel's Percentage Price Oscillator (PPO).
//
// PPO is calculated by subtracting the slower moving average from the faster moving
// average and then dividing the result by the slower moving average, expressed as a percentage.
//
//	PPO = 100 * (fast_ma - slow_ma) / slow_ma
//
// Reference:
//
// Appel, Gerald (2005). Technical Analysis: Power Tools for Active Investors.
type PercentagePriceOscillator struct {
	mu sync.RWMutex
	core.LineIndicator
	fastMA lineUpdater
	slowMA lineUpdater
	value  float64
	primed bool
}

// NewPercentagePriceOscillator returns an instance of the indicator created using supplied parameters.
func NewPercentagePriceOscillator(p *PercentagePriceOscillatorParams) (*PercentagePriceOscillator, error) {
	const (
		invalid   = "invalid percentage price oscillator parameters"
		fmts      = "%s: %s"
		fmtw      = "%s: %w"
		minLength = 2
	)

	if p.FastLength < minLength {
		return nil, fmt.Errorf(fmts, invalid, "fast length should be greater than 1")
	}

	if p.SlowLength < minLength {
		return nil, fmt.Errorf(fmts, invalid, "slow length should be greater than 1")
	}

	bc := p.BarComponent
	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	qc := p.QuoteComponent
	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	tc := p.TradeComponent
	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	var fastMA, slowMA lineUpdater

	var maLabel string

	switch p.MovingAverageType {
	case EMA:
		maLabel = "EMA"

		fast, e := exponentialmovingaverage.NewExponentialMovingAverageLength(
			&exponentialmovingaverage.ExponentialMovingAverageLengthParams{
				Length:         p.FastLength,
				FirstIsAverage: p.FirstIsAverage,
			})
		if e != nil {
			return nil, fmt.Errorf(fmtw, invalid, e)
		}

		slow, e := exponentialmovingaverage.NewExponentialMovingAverageLength(
			&exponentialmovingaverage.ExponentialMovingAverageLengthParams{
				Length:         p.SlowLength,
				FirstIsAverage: p.FirstIsAverage,
			})
		if e != nil {
			return nil, fmt.Errorf(fmtw, invalid, e)
		}

		fastMA = fast
		slowMA = slow
	default:
		maLabel = "SMA"

		fast, e := simplemovingaverage.NewSimpleMovingAverage(
			&simplemovingaverage.SimpleMovingAverageParams{Length: p.FastLength})
		if e != nil {
			return nil, fmt.Errorf(fmtw, invalid, e)
		}

		slow, e := simplemovingaverage.NewSimpleMovingAverage(
			&simplemovingaverage.SimpleMovingAverageParams{Length: p.SlowLength})
		if e != nil {
			return nil, fmt.Errorf(fmtw, invalid, e)
		}

		fastMA = fast
		slowMA = slow
	}

	mnemonic := fmt.Sprintf("ppo(%s%d/%s%d%s)", maLabel, p.FastLength, maLabel, p.SlowLength,
		core.ComponentTripleMnemonic(bc, qc, tc))
	desc := "Percentage Price Oscillator " + mnemonic

	ppo := &PercentagePriceOscillator{
		fastMA: fastMA,
		slowMA: slowMA,
		value:  math.NaN(),
	}

	ppo.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, ppo.Update)

	return ppo, nil
}

// IsPrimed indicates whether the indicator is primed.
func (s *PercentagePriceOscillator) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes the output data of the indicator.
func (s *PercentagePriceOscillator) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.PercentagePriceOscillator,
		s.LineIndicator.Mnemonic,
		s.LineIndicator.Description,
		[]core.OutputText{
			{Mnemonic: s.LineIndicator.Mnemonic, Description: s.LineIndicator.Description},
		},
	)
}

// Update updates the value of the indicator given the next sample.
func (s *PercentagePriceOscillator) Update(sample float64) float64 {
	const epsilon = 1e-8

	if math.IsNaN(sample) {
		return sample
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	slow := s.slowMA.Update(sample)
	fast := s.fastMA.Update(sample)
	s.primed = s.slowMA.IsPrimed() && s.fastMA.IsPrimed()

	if math.IsNaN(fast) || math.IsNaN(slow) {
		s.value = math.NaN()

		return s.value
	}

	if math.Abs(slow) < epsilon {
		s.value = 0
	} else {
		s.value = 100 * (fast - slow) / slow
	}

	return s.value
}

File: ts/indicators/gerald-appel/percentage-price-oscillator/percentage-price-oscillator.ts
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorType } from '../../core/indicator-type';
import { LineIndicator } from '../../core/line-indicator';
import { OutputType } from '../../core/outputs/output-type';
import { ExponentialMovingAverage } from '../../common/exponential-moving-average/exponential-moving-average';
import { SimpleMovingAverage } from '../../common/simple-moving-average/simple-moving-average';
import { PercentagePriceOscillatorOutput } from './percentage-price-oscillator-output';
import { PercentagePriceOscillatorParams, MovingAverageType } from './percentage-price-oscillator-params';

/** Interface for an indicator that accepts a scalar and returns a value. */
interface LineUpdater {
  update(sample: number): number;
  isPrimed(): boolean;
}

/** Function to calculate mnemonic of a __PercentagePriceOscillator__ indicator. */
export const percentagePriceOscillatorMnemonic = (params: PercentagePriceOscillatorParams): string => {
  const cm = componentTripleMnemonic(
    params.barComponent,
    params.quoteComponent,
    params.tradeComponent,
  );

  const maLabel = params.movingAverageType === MovingAverageType.EMA ? 'EMA' : 'SMA';

  return `ppo(${maLabel}${params.fastLength}/${maLabel}${params.slowLength}${cm})`;
};

/**
 * PercentagePriceOscillator is Gerald Appel's Percentage Price Oscillator (PPO).
 *
 * PPO is calculated by subtracting the slower moving average from the faster moving
 * average and then dividing the result by the slower moving average, expressed as a percentage.
 *
 *   PPO = 100 * (fast_ma - slow_ma) / slow_ma
 *
 * Reference:
 *
 * Appel, Gerald (2005). Technical Analysis: Power Tools for Active Investors.
 */
export class PercentagePriceOscillator extends LineIndicator {
  private fastMA: LineUpdater;
  private slowMA: LineUpdater;
  private value: number;

  /**
   * Constructs an instance given the parameters.
   */
  public constructor(params: PercentagePriceOscillatorParams) {
    super();

    const fastLength = Math.floor(params.fastLength);
    const slowLength = Math.floor(params.slowLength);

    if (fastLength < 2) {
      throw new Error('fast length should be greater than 1');
    }

    if (slowLength < 2) {
      throw new Error('slow length should be greater than 1');
    }

    this.mnemonic = percentagePriceOscillatorMnemonic(params);
    this.description = 'Percentage Price Oscillator ' + this.mnemonic;
    this.barComponent = params.barComponent;
    this.quoteComponent = params.quoteComponent;
    this.tradeComponent = params.tradeComponent;

    if (params.movingAverageType === MovingAverageType.EMA) {
      this.fastMA = new ExponentialMovingAverage({
        length: fastLength,
        firstIsAverage: params.firstIsAverage ?? false,
      });
      this.slowMA = new ExponentialMovingAverage({
        length: slowLength,
        firstIsAverage: params.firstIsAverage ?? false,
      });
    } else {
      this.fastMA = new SimpleMovingAverage({ length: fastLength });
      this.slowMA = new SimpleMovingAverage({ length: slowLength });
    }

    this.value = Number.NaN;
    this.primed = false;
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return {
      type: IndicatorType.PercentagePriceOscillator,
      mnemonic: this.mnemonic,
      description: this.description,
      outputs: [{
        kind: PercentagePriceOscillatorOutput.PercentagePriceOscillatorValue,
        type: OutputType.Scalar,
        mnemonic: this.mnemonic,
        description: this.description,
      }],
    };
  }

  /** Updates the value of the indicator given the next sample. */
  public update(sample: number): number {
    const epsilon = 1e-8;

    if (Number.isNaN(sample)) {
      return sample;
    }

    const slow = this.slowMA.update(sample);
    const fast = this.fastMA.update(sample);
    this.primed = this.slowMA.isPrimed() && this.fastMA.isPrimed();

    if (Number.isNaN(fast) || Number.isNaN(slow)) {
      this.value = Number.NaN;
      return this.value;
    }

    if (Math.abs(slow) < epsilon) {
      this.value = 0;
    } else {
      this.value = 100 * (fast - slow) / slow;
    }

    return this.value;
  }
}
